staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/capmanagers_test.go

```diff
@@ -0,0 +1,274 @@
+	}
+
+	if e, a := "fieldmanager_test_update_4", f.ManagedFields()[2].Manager; e != a {
+		t.Fatalf("exected third manager name to be %v, but got %v: %#v", e, a, f.ManagedFields())
+	}
+
+	expectManagesField(t, f, "ancient-changes", fieldpath.MakePathOrDie("one"))
+	expectManagesField(t, f, "ancient-changes", fieldpath.MakePathOrDie("two"))
+	expectManagesField(t, f, "fieldmanager_test_update_3", fieldpath.MakePathOrDie("three"))
+	expectManagesField(t, f, "fieldmanager_test_update_4", fieldpath.MakePathOrDie("four"))
+}
+
+func TestCapUpdateManagers(t *testing.T) {
+	set := func(fields ...string) *metav1.FieldsV1 {
+		s := fieldpath.NewSet()
+		for _, f := range fields {
+			s.Insert(fieldpath.MakePathOrDie(f))
+		}
+		b, err := s.ToJSON()
+		if err != nil {
+			panic(fmt.Sprintf("error building ManagedFieldsEntry for test: %v", err))
+		}
+		return &metav1.FieldsV1{Raw: b}
+	}
+	entry := func(name string, version string, order int, fields *metav1.FieldsV1) metav1.ManagedFieldsEntry {
+		return metav1.ManagedFieldsEntry{
+			Manager:    name,
+			APIVersion: version,
+			Operation:  "Update",
+			FieldsType: "FieldsV1",
+			FieldsV1:   fields,
+			Time:       &metav1.Time{Time: time.Time{}.Add(time.Hour * time.Duration(order))},
+		}
+	}
+
+	testCases := []struct {
+		name     string
+		input    []metav1.ManagedFieldsEntry
+		expected []metav1.ManagedFieldsEntry
+	}{
+		{
+			name: "one version, no ancient changes",
+			input: []metav1.ManagedFieldsEntry{
+				entry("update-manager1", "v1", 1, set("a")),
+				entry("update-manager2", "v1", 2, set("b")),
+				entry("update-manager3", "v1", 3, set("c")),
+				entry("update-manager4", "v1", 4, set("d")),
+			},
+			expected: []metav1.ManagedFieldsEntry{
+				entry("ancient-changes", "v1", 2, set("a", "b")),
+				entry("update-manager3", "v1", 3, set("c")),
+				entry("update-manager4", "v1", 4, set("d")),
+			},
+		}, {
+			name: "one version, one ancient changes",
+			input: []metav1.ManagedFieldsEntry{
+				entry("ancient-changes", "v1", 2, set("a", "b")),
+				entry("update-manager3", "v1", 3, set("c")),
+				entry("update-manager4", "v1", 4, set("d")),
+				entry("update-manager5", "v1", 5, set("e")),
+			},
+			expected: []metav1.ManagedFieldsEntry{
+				entry("ancient-changes", "v1", 3, set("a", "b", "c")),
+				entry("update-manager4", "v1", 4, set("d")),
+				entry("update-manager5", "v1", 5, set("e")),
+			},
+		}, {
+			name: "two versions, no ancient changes",
+			input: []metav1.ManagedFieldsEntry{
+				entry("update-manager1", "v1", 1, set("a")),
+				entry("update-manager2", "v2", 2, set("b")),
+				entry("update-manager3", "v1", 3, set("c")),
+				entry("update-manager4", "v1", 4, set("d")),
+				entry("update-manager5", "v1", 5, set("e")),
+			},
+			expected: []metav1.ManagedFieldsEntry{
+				entry("update-manager2", "v2", 2, set("b")),
+				entry("ancient-changes", "v1", 4, set("a", "c", "d")),
+				entry("update-manager5", "v1", 5, set("e")),
+			},
```

Comment:
In this test case, update-manager2 is not merged into a bucket manager and 3 and 4 are, even though it is older than 3 and 4, since merging 2 wouldn't make the list any smaller.

Reply:
Might make a good comment :)